File: lib/features/club/utils/coach_permissions.dart
import 'package:viro_team_v2/constants/firestore_fields.dart';

/// Droits coach configurables sur `clubs/{clubId}.coachPermissions`.
class CoachPermissions {
  const CoachPermissions({
    this.canCreateEvents = true,
    this.canManageTeamRoster = true,
    this.canInvitePlayers = true,
    this.canTakeAttendance = true,
    this.canViewFees = false,
  });

  final bool canCreateEvents;
  final bool canManageTeamRoster;
  final bool canInvitePlayers;
  final bool canTakeAttendance;
  final bool canViewFees;

  /// Defaults si le map est absent (clubs existants).
  static const CoachPermissions defaults = CoachPermissions();

  /// Parse le map Firestore avec defaults.
  factory CoachPermissions.fromMap(Map<String, dynamic>? map) {
    if (map == null) return defaults;
    return CoachPermissions(
      canCreateEvents: map[FirestoreFields.canCreateEvents] as bool? ?? true,
      canManageTeamRoster:
          map[FirestoreFields.canManageTeamRoster] as bool? ?? true,
      canInvitePlayers: map[FirestoreFields.canInvitePlayers] as bool? ?? true,
      canTakeAttendance: map[FirestoreFields.canTakeAttendance] as bool? ?? true,
      canViewFees: map[FirestoreFields.canViewFees] as bool? ?? false,
    );
  }

  /// True si le viewer coach peut créer des événements.
  bool allowsCreateEvents({required bool isAdmin, required bool isCoach}) {
    if (isAdmin) return true;
    return isCoach && canCreateEvents;
  }

  /// True si le viewer coach peut gérer le roster de ses équipes.
  bool allowsManageTeamRoster({required bool isAdmin, required bool isCoach}) {
    if (isAdmin) return true;
    return isCoach && canManageTeamRoster;
  }

  /// True si le viewer coach peut inviter / ajouter des joueurs.
  bool allowsInvitePlayers({required bool isAdmin, required bool isCoach}) {
    if (isAdmin) return true;
    return isCoach && canInvitePlayers;
  }

  /// True si le viewer coach peut prendre les présences.
  bool allowsTakeAttendance({required bool isAdmin, required bool isCoach}) {
    if (isAdmin) return true;
    return isCoach && canTakeAttendance;
  }

  /// True si le viewer coach peut voir les cotisations (lecture).
  bool allowsViewFees({required bool isAdmin, required bool isCoach}) {
    if (isAdmin) return true;
    return isCoach && canViewFees;
  }

  /// Payload Firestore pour `clubs/{clubId}.coachPermissions`.
  Map<String, bool> toMap() => {
        FirestoreFields.canCreateEvents: canCreateEvents,
        FirestoreFields.canManageTeamRoster: canManageTeamRoster,
        FirestoreFields.canInvitePlayers: canInvitePlayers,
        FirestoreFields.canTakeAttendance: canTakeAttendance,
        FirestoreFields.canViewFees: canViewFees,
      };

  CoachPermissions copyWith({
    bool? canCreateEvents,
    bool? canManageTeamRoster,
    bool? canInvitePlayers,
    bool? canTakeAttendance,
    bool? canViewFees,
  }) {
    return CoachPermissions(
      canCreateEvents: canCreateEvents ?? this.canCreateEvents,
      canManageTeamRoster:
          canManageTeamRoster ?? this.canManageTeamRoster,
      canInvitePlayers: canInvitePlayers ?? this.canInvitePlayers,
      canTakeAttendance: canTakeAttendance ?? this.canTakeAttendance,
      canViewFees: canViewFees ?? this.canViewFees,
    );
  }
}

File: portal/src/lib/auth/coachPermissions.ts
import { Fields } from "@/lib/firebase/constants";

/** Droits coach configurables sur le doc club. */
export type CoachPermissions = {
  canCreateEvents: boolean;
  canManageTeamRoster: boolean;
  canInvitePlayers: boolean;
  canTakeAttendance: boolean;
  canViewFees: boolean;
};

/** Defaults si le map est absent (clubs existants). */
export const DEFAULT_COACH_PERMISSIONS: CoachPermissions = {
  canCreateEvents: true,
  canManageTeamRoster: true,
  canInvitePlayers: true,
  canTakeAttendance: true,
  canViewFees: false,
};

/** Parse `clubs/{id}.coachPermissions` avec defaults. */
export function parseCoachPermissions(raw: unknown): CoachPermissions {
  const map =
    raw && typeof raw === "object" && !Array.isArray(raw)
      ? (raw as Record<string, unknown>)
      : {};
  return {
    canCreateEvents:
      map[Fields.canCreateEvents] === undefined
        ? DEFAULT_COACH_PERMISSIONS.canCreateEvents
        : Boolean(map[Fields.canCreateEvents]),
    canManageTeamRoster:
      map[Fields.canManageTeamRoster] === undefined
        ? DEFAULT_COACH_PERMISSIONS.canManageTeamRoster
        : Boolean(map[Fields.canManageTeamRoster]),
    canInvitePlayers:
      map[Fields.canInvitePlayers] === undefined
        ? DEFAULT_COACH_PERMISSIONS.canInvitePlayers
        : Boolean(map[Fields.canInvitePlayers]),
    canTakeAttendance:
      map[Fields.canTakeAttendance] === undefined
        ? DEFAULT_COACH_PERMISSIONS.canTakeAttendance
        : Boolean(map[Fields.canTakeAttendance]),
    canViewFees:
      map[Fields.canViewFees] === undefined
        ? DEFAULT_COACH_PERMISSIONS.canViewFees
        : Boolean(map[Fields.canViewFees]),
  };
}

/** Payload Firestore pour coachPermissions. */
export function coachPermissionsToFirestore(
  permissions: CoachPermissions,
): Record<string, boolean> {
  return {
    [Fields.canCreateEvents]: permissions.canCreateEvents,
    [Fields.canManageTeamRoster]: permissions.canManageTeamRoster,
    [Fields.canInvitePlayers]: permissions.canInvitePlayers,
    [Fields.canTakeAttendance]: permissions.canTakeAttendance,
    [Fields.canViewFees]: permissions.canViewFees,
  };
}

/** Libellés FR des toggles paramètres. */
export const COACH_PERMISSION_LABELS: {
  key: keyof CoachPermissions;
  label: string;
  description: string;
}[] = [
  {
    key: "canCreateEvents",
    label: "Créer des événements",
    description: "Le coach peut créer et gérer le planning de ses équipes.",
  },
  {
    key: "canManageTeamRoster",
    label: "Gérer le roster",
    description: "Ajouter des joueurs aux équipes qu’il entraîne.",
  },
  {
    key: "canInvitePlayers",
    label: "Inviter des joueurs",
    description: "Créer une fiche membre et envoyer une invitation.",
  },
  {
    key: "canTakeAttendance",
    label: "Prendre les présences",
    description: "Réserver pour une future UI présences (flag stocké).",
  },
  {
    key: "canViewFees",
    label: "Voir les cotisations",
    description: "Accès lecture au suivi des cotisations du club.",
  },
];
